front/src/HOC/withLogger.test.js

```diff
@@ -0,0 +1,211 @@
+      1,
+      'component mounted'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      2,
+      'component should update'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      3,
+      'component updated'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      4,
+      'component will unmount'
+    );
+  });
+  test('should add log function with specified HOC name', () => {
+    const WrappedComponent = withLogger()(TestComponent);
+    expect(WrappedComponent.displayName).toBe(
+      'withLogger(TestComponent)'
+    );
+  });
+  test('should log specified lifecycles', () => {
+    const WrappedComponent = withLogger({
+      output: fakeLogger,
+      lifecycles: ['didMount', 'willUnmount'],
+    })(TestComponent);
+    const { unmount } = render(<WrappedComponent />);
+    expect(fakeLogger).toHaveBeenCalledTimes(1);
+    unmount(<WrappedComponent />);
+    expect(fakeLogger).toHaveBeenCalledTimes(2);
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      1,
+      'component mounted'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      2,
+      'component will unmount'
+    );
+  });
+  test('should log props', () => {
+    const WrappedComponent = withLogger({
+      output: fakeLogger,
+      lifecycles: ['props'],
+    })(TestComponent);
+    const {} = render(<WrappedComponent q="1" w="erty" />);
+    expect(fakeLogger).toHaveBeenCalledTimes(1);
+    expect(fakeLogger).toHaveBeenNthCalledWith(1, {
+      q: '1',
+      w: 'erty',
+    });
+  });
+  test('should log props first', () => {
+    const WrappedComponent = withLogger({
+      output: fakeLogger,
+      lifecycles: [
+        'props',
+        'didMount',
+        'shouldUpdate',
+        'willUnmount',
+      ],
+    })(TestComponent);
+    const { rerender, unmount } = render(
+      <WrappedComponent q="1" w="erty" />
+    );
+    expect(fakeLogger).toHaveBeenCalledTimes(2);
+    rerender(<WrappedComponent qw="1" />);
+    expect(fakeLogger).toHaveBeenCalledTimes(4);
+    unmount(<WrappedComponent />);
+    expect(fakeLogger).toHaveBeenCalledTimes(5);
+    expect(fakeLogger).toHaveBeenNthCalledWith(1, {
+      q: '1',
+      w: 'erty',
+    });
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      2,
+      'component mounted'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(3, {
+      qw: '1',
+    });
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      4,
+      'component should update'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      5,
+      'component will unmount'
+    );
+  });
+  test('should add log function with specified params for class component', () => {
+    const WrappedComponent = withLogger({
+      output: fakeLogger,
+      lifecycles: [
+        'didMount',
+        'shouldUpdate',
+        'didUpdate',
+        'willUnmount',
+      ],
+    })(TestClassComponent());
+    const { rerender, unmount } = render(
+      <WrappedComponent log={fakeLogger} />
```

Comment:
What should happen if we pass here a different logger, e.g. `const fakeLogger2 = jest.fn()` ?

Reply:
mock data will be differ for parent (wrapped with fakeLogger2) and child (to wrap with fakeLogger) component. We won't be able to recognize the order of calls.